core/lib/zksync_core/src/state_keeper/keeper.rs

```diff
@@ -234,6 +226,59 @@ impl ZkSyncStateKeeper {
         Err(Error::Canceled)
     }
 
+    async fn load_protocol_upgrade_tx(
+        &mut self,
+        pending_miniblocks: &[MiniblockExecutionData],
+        protocol_version: ProtocolVersionId,
+        first_batch_in_shared_bridge: bool,
```

Comment:
I don't like that this upgrade logic is spread in 3 places (one here, one in line 140 - before the loop, and one in 220 - at the end of the loop).

could we instead move the logic to this function, and remove stuff from line 220 and 140?

Reply:
on line 220 we're calling `self.io.load_upgrade_tx` in a loop - since the state keeper has to process upgrade transactions routinely, not only at startup. Therefore since it does not have to perform additional checks or waiting loops, calling `self.load_protocol_upgrade_tx` is not really appropriate.

I don't see how it's possible to incorporate this logic into the existing function.

As for line 140, would you propose passing the L1BatchNumber into the function instead of `first_batch_in_shared_bridge`?

Reply:
yes, we could put l1 batch number  (as argument here).